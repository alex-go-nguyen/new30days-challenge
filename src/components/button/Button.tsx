import React from 'react';
import styles from './Button.module.scss';
import classNames from 'classnames';

type ButtonVariant = 'primary' | 'secondary';

interface ButtonProps {
    /**
     * What color of the button's background?
     */
    backgroundColor?: string;
    /**
     * Button contents
     */
    label: string;
    /**
     * Choose hoverColor
     */
    hoverColor?: string;
    /**
     * Active or disable button
     */
    disabled?: boolean;
    /**
     * Set up onclick
     */
    onClick?: () => void;

    variant?: ButtonVariant;
}

export default function Button({
    variant = 'primary',
    label,
    backgroundColor,
    hoverColor,
    disabled,
    onClick,
}: ButtonProps) {
    return (
        <button
            type="button"
            className={classNames({
                [styles[variant]]: true,
                [styles.disabled]: disabled,
                [styles.button]: true,
            })}
            onClick={onClick}
        >
            {label}
            <style jsx>{`
                button {
                    background-color: ${backgroundColor};
                }
                button:hover {
                    background-color: ${hoverColor};
                }
            `}</style>
        </button>
    );
}
