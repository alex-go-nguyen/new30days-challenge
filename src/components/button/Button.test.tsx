import React from 'react';
import renderer, { ReactTestRendererJSON, act } from 'react-test-renderer';
import Button from './Button';

describe('Button component', () => {
    test('should render correctly', () => {
        const component = renderer.create(
            <Button color="success" variant="primary">
                Button
            </Button>,
        );
        let tree = component.toJSON();
        expect(tree).toMatchSnapshot();
    });

    test('triggers onClick event', () => {
        const mockOnClick = jest.fn();
        const component = renderer.create(<Button onClick={mockOnClick}>Button</Button>);
        act(() => component.root.props.onClick());
        expect(mockOnClick).toHaveBeenCalledTimes(1);
    });

    test('Should have correct content', () => {
        const component = renderer.create(<Button>Click me!</Button>);
        const tree = component.toJSON() as ReactTestRendererJSON;
        expect(tree.children).toEqual(['Click me!']);
    });
});
