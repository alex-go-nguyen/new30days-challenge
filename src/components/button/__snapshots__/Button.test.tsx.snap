// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Button component should render correctly 1`] = `
<button
  className="primary button md success"
  type="button"
>
  Button
</button>
`;
